### PR TITLE
Hackathon! Short cycle detection algorithm

### PR DESCRIPTION
**Pull request by wpride**
Vastly improves the error message for cyclical references in forms

cross-request: https://github.com/dimagi/formplayer/pull/547/files

### CODE REVIEW

File: src/main/java/org/javarosa/xform/parse/XFormParser.java [hunk 1]
```diff
@@ -2898,14 +2895,9 @@ private void checkDependencyCycles() {
...
-            String errorMessage = "Logic Dependency Cycle:\n";
-            for (int i = 0; i < edges.size(); i++) {
-                TreeReference[] edge = (TreeReference[])edges.elementAt(i);
-                errorMessage += edge[0].toString() + " references " + edge[1].toString() + "\n";
-            }
-            reporter.error(errorMessage);
-
-            throw new RuntimeException("Logic is cyclical, referencing itself.");
+            ArrayList<String> cycleString = new ShortestCycleAlgorithm(edges).getCycle();
+            reporter.error("Logic is cyclical, referencing itself. Shortest Cycle " + cycleString);
```
Comment (ctsims) [inline src/main/java/org/javarosa/xform/parse/XFormParser.java]: Relying on the toString of ArrayList here doesn't seem safe to me. I think we need a deliberate representation of what this should look like

Comment (ctsims) [reply src/main/java/org/javarosa/xform/parse/XFormParser.java]: also: this message is duplicated and should be put in an intermediate String and re-used.

**Comment by ctsims** [comment]
@wpride One other thing: The cycle message shows up in two different contexts in the context Ticket from jessica. Does this message replace both of those places?

**Comment by wpride** [comment]
@ctsims thanks, updated

**Comment by amstone326** [comment]
@wpride Why does this have to target the formplayer branch instead of just master?

**Comment by wpride** [comment]
@amstone326 Because Formplayer is the primary mechanism delivering these parse error messages these days and I find it easier to develop on my Formplayer IDE than Android. Will get into `master` after this is merged

File: src/main/java/org/javarosa/xform/parse/XFormParser.java [hunk 2]
```diff
@@ -2057,7 +2054,7 @@ private void addMainInstanceToFormDef(Element e, FormInstance instanceModel) {
         //TreeElement root = buildInstanceStructure(e, null);
         loadInstanceData(e, instanceModel.getRoot());
 
-        checkDependencyCycles();
+        //checkDependencyCycles();
```
Comment (amstone326) [inline src/main/java/org/javarosa/xform/parse/XFormParser.java]: this shouldn't stay commented

**Comment by amstone326** [comment]
@wpride Sorry I'm still confused; what about this PR couldn't be targeted to master as-is?

**Comment by wpride** [comment]
@amstone326 thanks

@ctsims that reminds me - there are two places where we check for cyclical references, in `finalizeTriggerables` in `FormDef` and in `addMainInstanceToFormDef` in `XFormParser`. `finalizeTriggerables` is only claled in two places - once in `XFormParser` immediately after we call `checkDependencyCycles` and again in `readExternal` of `FormDef`. So unless we saved a `FormDef` with cyclical references (which shouldn't be possible?) I'm confused how we could end up in the second call path

**Comment by wpride** [comment]
@amstone326 It could be I just want to finish all of the changes/iteration in one PR rather than two

**Comment by dannyroberts** [comment]
@wpride @ctsims Is this good to go?

**Comment by wpride** [comment]
This is gtg from my perspective - @ctsims

**Comment by wpride** [comment]
Merging since this is already merged in `master`

**Pull request merged**